pkg/storaged/client.js

```diff
@@ -206,6 +214,203 @@ function is_multipath_master(block) {
+        debug("btrfs_pol new subvols:", uuids_subvols);
+        client.uuids_btrfs_subvols = uuids_subvols;
+        client.uuids_btrfs_usage = uuids_usage;
+        debug("btrfs_pol usage:", uuids_usage);
+        client.uuids_btrfs_default_subvol = btrfs_default_subvol;
+        debug("btrfs_pol default subvolumes:", btrfs_default_subvol);
+        client.update();
+    }
+}
+
+function btrfs_findmnt_poll() {
+    if (!client.btrfs_mounts)
+        client.btrfs_mounts = { };
+
+    const update_btrfs_mounts = output => {
+        const btrfs_mounts = {};
+        try {
+            // Extract the data into a { uuid: { subvolid: { subvol, target } } }
+            const mounts = JSON.parse(output);
+            if ("filesystems" in mounts) {
+                for (const fs of mounts.filesystems) {
+                    const subvolid_match = fs.options.match(/subvolid=(?<subvolid>\d+)/);
+                    const subvol_match = fs.options.match(/subvol=(?<subvol>[\w\\/]+)/);
+                    if (!subvolid_match && !subvol_match) {
+                        console.warn("findmnt entry without subvol and subvolid", fs);
+                        break;
+                    }
+
+                    const { subvolid } = subvolid_match.groups;
+                    const { subvol } = subvol_match.groups;
+                    const subvolume = {
+                        pathname: subvol,
+                        id: subvolid,
+                        mount_points: [fs.target],
+                    };
+
+                    if (!(fs.uuid in btrfs_mounts)) {
+                        btrfs_mounts[fs.uuid] = { };
+                    }
+
+                    // We need to handle multiple mounts, they are listed seperate.
+                    if (subvolid in btrfs_mounts[fs.uuid]) {
+                        btrfs_mounts[fs.uuid][subvolid].mount_points.push(fs.target);
+                    } else {
+                        btrfs_mounts[fs.uuid][subvolid] = subvolume;
+                    }
+                }
+            }
+        } catch (exc) {
+            if (exc.message)
+                console.error("unable to parse findmnt JSON output", exc);
+        }
+
+        // Update client state
+        if (!deep_equal(client.btrfs_mounts, btrfs_mounts)) {
+            client.btrfs_mounts = btrfs_mounts;
+            debug("btrfs_findmnt_poll mounts:", client.btrfs_mounts);
+            client.update();
+        }
+    };
+
+    const findmnt_poll = () => {
+        return cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "--poll"], { superuser: true, err: "message" }).stream(() => {
+            cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "-o", "UUID,OPTIONS,TARGET", "--json"],
+                          { superuser: true, err: "message" }).then(output => update_btrfs_mounts(output)).catch(err => {
+                // When there are no btrfs filesystems left this can fail and thus we need to manually reset the mount info.
```

Comment:
Will the poll ever be restarted if a new btrfs shows up?

Reply:
So, on Ubuntu does not have btrfs at all, so this should work

Reply:
What? What do you mean Ubuntu does not have btrfs at all?

Reply:
It has, but when we first open the Cockpit page there would be no btrfs filesystem. I thought we created the btrfs filesystem after loading Cockpit but actually the tests first create the fs. So let me verify on a bare ubuntu test image should be easy enough.

Reply:
Verified in a VM that `findmnt` does the right thing:

```
root@ubuntu:~# findmnt --type btrfs --mtab --poll
ACTION     TARGET
                SOURCE   FSTYPE OPTIONS
mount      /mnt /dev/sda btrfs  rw,relatime,discard=async,space_cache=v2,subvolid=5,subvol=/

────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
SSD detected:       no
Zoned device:       no
Incompat features:  extref, skinny-metadata, no-holes, free-space-tree
Runtime features:   free-space-tree
Checksum:           crc32c
Number of devices:  1
Devices:
   ID        SIZE  PATH
    1    10.00GiB  /dev/sda

root@ubuntu:~# mount /dev/sda /mnt/
```

all the lines are the tmux separator :)

Reply:
> What? What do you mean Ubuntu does not have btrfs at all?

Our CI test images don't have a btrfs filesystem initially. I am sure Ubuntu in general supports btrfs, and we run all the tests on it.

Reply:
> Verified in a VM that findmnt does the right thing:

I meant, if the `findmnt --poll` ever exits for any reason, will we restart it? I don't see the place where we do.
And if we don't, mount point monitoring is going to be broken from then on, no?